src/Control/Monad/Throws.hs

```diff
@@ -0,0 +1,104 @@
+{-# language FunctionalDependencies #-}
+{-# OPTIONS_GHC -Wredundant-constraints #-}
+
+module Control.Monad.Throws where
+
+import Control.Monad.Catch (MonadThrow, MonadCatch, throwM, Exception)
+import Lib
+import GHC.Stack
+import GHC.TypeLits
+import qualified Control.Exception.Safe as Exception
+
+class (Monad m, Exception err) => MonadThrows err m | m -> err where
+    throwAllChecked :: (HasCallStack) => err -> m a
+
+throwChecked :: (MonadThrows err m, ExceptionSubtype err e) => e -> m a
+throwChecked = throwAllChecked . inject
+
+instance
+    ( Exception err, MonadThrow m
```

Comment:
Hmmm, this is bad.

We want to be able to say something like:

```haskell
instance (MonadThrow m, SomeException :< err) => MonadThrow (CheckedT err m)
```

That is, "if you want to call `throwM` to throw an unchecked exception in a `CheckedT` action, then you must incur the unchecked nature by pushing `SomeException :< err`."

But if we require `MonadThrow m => MonadThrows err (CheckedT err m)`, then we're saying: if you want to throw a *checked* exception, then we necessarily incur (and provide!) the ability to throw unchecked exceptions.

That's precisely what we want to avoid.

We could go for the `ExceptT` approach - then we only need `Monad`. But this incurs the performance problem which we're trying to avoid.

Another option is to use an explicit capability:

```haskell
newtype CheckedT err m a 
    = CheckedT (ReaderT (Throw m) m a)

newtype Throw m = MkThrow { runThrow :: forall e a. (Exception e) => e -> m a }
```

I'd want to verify that the callstack behavior works like I want it to, but I suspect it should. This may also make performance worse - benchmarking `ReaderT (Throw m) m a` vs `ExceptT e m a` would be a useful experiment.

Another option is to use [`Control.Exception.throw :: (Exception e) => e -> a`](https://www.stackage.org/haddock/lts-23.1/base-4.19.2.0/Control-Exception.html#v:throw). But this would hardcode throwing behavior, meaning you couldn't use [`Control.Exception.Annotated.throwWithCallStack`](https://hackage.haskell.org/package/annotated-exception-0.3.0.2/docs/Control-Exception-Annotated.html#v:throwWithCallStack). So I think that's also dead in the water.